pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1037 @@
+		queue:                      workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "trafficControlGroup"),
+	}
+	c.trafficControlInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addTC,
+			UpdateFunc: c.updateTC,
+			DeleteFunc: c.deleteTC,
+		},
+		resyncPeriod,
+	)
+	c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addPod,
+			UpdateFunc: c.updatePod,
+			DeleteFunc: c.deletePod,
+		},
+		resyncPeriod,
+	)
+	c.namespaceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addNamespace,
+			UpdateFunc: c.updateNamespace,
+			DeleteFunc: nil,
+		},
+		resyncPeriod,
+	)
+	podUpdateSubscriber.Subscribe(c.processPodUpdate)
+	return c
+}
+
+func tcInfoKeyFunc(obj interface{}) (string, error) {
+	return obj.(*trafficControlState).name, nil
+}
+
+func tcTargetPortIndexFunc(obj interface{}) ([]string, error) {
+	tcState := obj.(*trafficControlState)
+	return []string{strconv.Itoa(int(tcState.targetPort))}, nil
+}
+
+func tcReturnPortIndexFunc(obj interface{}) ([]string, error) {
+	tcState := obj.(*trafficControlState)
+	return []string{strconv.Itoa(int(tcState.returnPort))}, nil
+}
+
+// processPodUpdate will be called when CNIServer publishes a Pod update event. It triggers the event for the effective
+// TrafficControl of the Pod.
+func (c *Controller) processPodUpdate(e interface{}) {
+	c.podToTCBindingsMutex.RLock()
+	defer c.podToTCBindingsMutex.RUnlock()
+	podEvent := e.(types.PodUpdate)
+	pod := k8s.NamespacedName(podEvent.PodNamespace, podEvent.PodName)
+	binding, exists := c.podToTCBindings[pod]
+	if !exists {
+		return
+	}
+	c.queue.Add(binding.effectiveTC)
+}
+
+func (c *Controller) matchedPod(pod *v1.Pod, to *v1alpha2.AppliedTo) bool {
+	if to.NamespaceSelector == nil && to.PodSelector == nil {
+		return false
+	}
+	if to.NamespaceSelector != nil {
+		namespace, _ := c.namespaceLister.Get(pod.Namespace)
+		if namespace == nil {
+			return false
+		}
+		nsSelector, _ := metav1.LabelSelectorAsSelector(to.NamespaceSelector)
+		if !nsSelector.Matches(labels.Set(namespace.Labels)) {
+			return false
+		}
+	}
+	if to.PodSelector != nil {
+		podSelector, _ := metav1.LabelSelectorAsSelector(to.PodSelector)
+		if !podSelector.Matches(labels.Set(pod.Labels)) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (c *Controller) filterAffectedTCsByPod(pod *v1.Pod) sets.String {
+	affectedTCs := sets.NewString()
+	allTCs, _ := c.trafficControlLister.List(labels.Everything())
+	for _, tc := range allTCs {
+		if c.matchedPod(pod, &tc.Spec.AppliedTo) {
+			affectedTCs.Insert(tc.GetName())
+		}
+	}
+	return affectedTCs
+}
+
+func (c *Controller) addPod(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	if pod.Spec.HostNetwork {
```

Comment:
Should we ignore Pods without an IP?

Reply:
It should be possible a Pod has no OVS port created yet, and then when the OVS port is ready we should handle TC again, but in updatePod we do not check that. @tnqn : thoughts?

Reply:
> but in updatePod we do not check that.
Yes, we don't check if the OVS port is ready in `updatePod`.

Reply:
Ok. The approach sounds good to me.